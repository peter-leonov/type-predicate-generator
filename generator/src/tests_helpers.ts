import ts from "typescript";
import { factory } from "typescript";
import assert from "assert";
import {
  ensureNoErrors,
  newVFSProgram,
  sourceFileToDeclarationSymbols,
} from "./compile";

export function compile(
  source: string
): [ts.TypeChecker, ts.Type, ts.Symbol] {
  const program = newVFSProgram(source);
  ensureNoErrors(program);

  const sourceFile = program.getSourceFiles()[1];
  assert(
    sourceFile,
    "the program must have 2 source files, one for the lib and one for the actual test code"
  );

  const checker = program.getTypeChecker();

  // This finds the LAST type definition in the source.
  let symbol = sourceFileToDeclarationSymbols(checker, sourceFile).at(
    -1
  );
  assert(
    symbol,
    "at least one symbol must be present in the test case"
  );
  const type = checker.getDeclaredTypeOfSymbol(symbol);
  return [checker, type, symbol];
}

export function printNodes(nodes: ts.Statement[]): string {
  const resultFile = factory.updateSourceFile(
    ts.createSourceFile(
      "guards.ts",
      "",
      ts.ScriptTarget.Latest,
      /*setParentNodes*/ false,
      ts.ScriptKind.TS
    ),
    nodes
  );

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });
  return printer.printFile(resultFile);
}
