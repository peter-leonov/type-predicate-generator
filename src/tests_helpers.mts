import ts from "typescript";
import { unimplemented } from "./helpers.mts";
import { ok } from "assert";

export function compile(
  source: string
): [ts.TypeChecker, ts.Type, ts.Symbol] {
  const fileName = "/source.ts";

  const sourceFiles = new Map<string, ts.SourceFile>();
  sourceFiles.set(
    fileName,
    ts.createSourceFile(fileName, source, ts.ScriptTarget.ESNext)
  );

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => sourceFiles.has(fileName),
    getSourceFile: (fileName) => sourceFiles.get(fileName),
    getDefaultLibFileName: () => "lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: unimplemented,
  };

  const program = ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
    },
    compilerHost
  );

  const sourceFile = program.getSourceFiles()[0];
  ok(sourceFile);

  const checker = program.getTypeChecker();

  let symbol: ts.Symbol | undefined;
  ts.forEachChild(sourceFile, (node) => {
    if (!ts.isTypeAliasDeclaration(node)) {
      return;
    }

    symbol = checker.getSymbolAtLocation(node.name);
    ok(symbol);
  });

  ok(symbol);
  const type = checker.getDeclaredTypeOfSymbol(symbol);
  return [checker, type, symbol];
}
