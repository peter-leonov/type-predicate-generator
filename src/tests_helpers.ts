import ts from "typescript";
import { factory } from "typescript";
import { unimplemented } from "./helpers";
import { ok } from "assert";

/**
 * This is required for both `X[]` and `Array<X>` to work properly.
 * Otherwise `checker.isArrayType()` is not gonna work.
 */
const libFileContent = `/// <reference no-default-lib="true"/>
interface Boolean {}
interface Function {}
interface CallableFunction {}
interface NewableFunction {}
interface IArguments {}
interface Number { toExponential: any; }
interface Object {}
interface RegExp {}
interface String { charAt: any; }
interface Array<T> { length: number; [n: number]: T; }
interface ReadonlyArray<T> {}
declare const console: { log(msg: any): void; };`;

export function compile(
  source: string
): [ts.TypeChecker, ts.Type, ts.Symbol] {
  const fileName = "/source.ts";

  const files = new Map<string, string>();
  files.set(fileName, source);

  const libFileName = "/lib/lib.d.ts";
  files.set(libFileName, libFileContent);

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => files.has(fileName),
    getSourceFile: (fileName, options) => {
      const content = files.get(fileName);
      ok(content);
      return ts.createSourceFile(fileName, content, options);
    },
    getDefaultLibFileName: () => "/lib/lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: (path) => files.get(path),
  };

  const program = ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
      strict: true,
      noEmit: true,
      isolatedModules: true,
      // don't autoload all the types from the node_modules/@types
      types: [],
    },
    compilerHost
  );

  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  if (allDiagnostics.length != 0) {
    allDiagnostics.forEach((diagnostic) => {
      var message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        "\n"
      );
      if (!diagnostic.file) {
        console.error(message);
        return;
      }
      var { line, character } =
        diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!
        );
      console.error(
        `${diagnostic.file.fileName} (${line + 1},${
          character + 1
        }): ${message}`
      );
    });

    throw new Error(
      "ts.getPreEmitDiagnostics() found some issues listed above"
    );
  }

  const sourceFile = program.getSourceFiles()[1];
  ok(sourceFile);

  const checker = program.getTypeChecker();

  // This finds the LAST type definition in the source.
  let symbol: ts.Symbol | undefined;
  ts.forEachChild(sourceFile, (node) => {
    if (!ts.isTypeAliasDeclaration(node)) {
      return;
    }

    symbol = checker.getSymbolAtLocation(node.name);
    ok(symbol);
    // console.dir(checker.getTypeFromTypeNode(node.type));
  });

  ok(symbol);
  const type = checker.getDeclaredTypeOfSymbol(symbol);
  return [checker, type, symbol];
}

export function printNodes(nodes: ts.Statement[]): string {
  const resultFile = factory.updateSourceFile(
    ts.createSourceFile(
      "guards.ts",
      "",
      ts.ScriptTarget.Latest,
      /*setParentNodes*/ false,
      ts.ScriptKind.TS
    ),
    nodes
  );

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });
  return printer.printFile(resultFile);
}
