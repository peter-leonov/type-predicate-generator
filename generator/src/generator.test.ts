import { expect, test } from "vitest";
import {
  TypeGuardGenerator,
  UnsupportedUnionMember,
} from "./generator.js";
import {
  ArrayType,
  LiteralType,
  ObjectType,
  PrimitiveType,
  AliasType,
  TypeModel,
  UnionType,
} from "./model.js";
import { printNodes } from "./tests_helpers.js";

export function generate(model: TypeModel): string {
  const tgg = new TypeGuardGenerator();
  tgg.addRootTypeGuardFor(model);
  return printNodes(tgg.getGuards());
}

test("undefined", () => {
  expect(
    generate(new LiteralType({ aliasName: "X" }, undefined))
  ).toMatchSnapshot();
});

test("null", () => {
  expect(
    generate(new LiteralType({ aliasName: "X" }, null))
  ).toMatchSnapshot();
});

test("union of primitive and literal types", () => {
  expect(
    generate(
      new UnionType({ aliasName: "Union" }, [
        new PrimitiveType({}, "string"),
        new PrimitiveType({}, "number"),
        new LiteralType({}, null),
      ])
    )
  ).toMatchSnapshot();
});

test("union of object types", () => {
  expect(() =>
    generate(
      new UnionType({ aliasName: "Union" }, [
        new ObjectType({}, { a: new PrimitiveType({}, "number") }),
        new ObjectType({}, { b: new PrimitiveType({}, "string") }),
      ])
    )
  ).toThrow(UnsupportedUnionMember);
});

test("object with primitive types", () => {
  expect(
    generate(
      new ObjectType(
        { aliasName: "X" },
        {
          a: new PrimitiveType({}, "number"),
          b: new PrimitiveType({}, "string"),
          c: new PrimitiveType({}, "boolean"),
        }
      )
    )
  ).toMatchSnapshot();
});

test("reference type in an object", () => {
  expect(
    generate(
      new ObjectType(
        { aliasName: "X" },
        {
          a: new AliasType({ aliasName: "A" }, "A"),
          b: new AliasType({ aliasName: "B" }, "B"),
        }
      )
    )
  ).toMatchSnapshot();
});

test("reference type in a union", () => {
  expect(
    generate(
      new UnionType({ aliasName: "X" }, [
        new PrimitiveType({}, "number"),
        new AliasType({ aliasName: "A" }, "A"),
      ])
    )
  ).toMatchSnapshot();
});

test("nested object", () => {
  expect(
    generate(
      new ObjectType(
        { aliasName: "X" },
        {
          a: new ObjectType(
            {},
            {
              b: new ObjectType(
                {},
                {
                  c: new PrimitiveType({}, "string"),
                }
              ),
            }
          ),
        }
      )
    )
  ).toMatchSnapshot();
});

test("array of a primitive type", () => {
  expect(
    generate(
      new ArrayType(
        { aliasName: "X" },
        new PrimitiveType({}, "number")
      )
    )
  ).toMatchSnapshot();
});

test("array of a primitive type union", () => {
  expect(
    generate(
      new ArrayType(
        { aliasName: "X" },
        new UnionType({}, [
          new PrimitiveType({}, "number"),
          new PrimitiveType({}, "string"),
        ])
      )
    )
  ).toMatchSnapshot();
});

test("array of an object type", () => {
  expect(
    generate(
      new ArrayType(
        { aliasName: "X" },
        new ObjectType(
          {},
          {
            a: new PrimitiveType({}, "number"),
            b: new PrimitiveType({}, "string"),
          }
        )
      )
    )
  ).toMatchSnapshot();
});

test("two arrays of a primitive type", () => {
  expect(
    generate(
      new ObjectType(
        { aliasName: "X" },
        {
          a: new ArrayType({}, new PrimitiveType({}, "number")),
          b: new ArrayType({}, new PrimitiveType({}, "string")),
        }
      )
    )
  ).toMatchSnapshot();
});
