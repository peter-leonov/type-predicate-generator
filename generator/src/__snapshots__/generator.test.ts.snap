// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array of a primitive type 1`] = `
"export function isX(root: unknown): root is X {
    type Element = X[number];
    function isElement(root: unknown): root is Element {
        if (!(typeof root === "number")) {
            return false;
        }
        ensureType<Element>(root);
        return true;
    }
    if (!(safeIsArray(root) && root.every(isElement))) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`array of a primitive type union 1`] = `
"export function isX(root: unknown): root is X {
    type Element = X[number];
    function isElement(root: unknown): root is Element {
        if (!((typeof root === "number") || (typeof root === "string"))) {
            return false;
        }
        ensureType<Element>(root);
        return true;
    }
    if (!(safeIsArray(root) && root.every(isElement))) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`array of an object type 1`] = `
"export function isX(root: unknown): root is X {
    type Element = X[number];
    function isElement(root: unknown): root is Element {
        if (!(typeof root === "object" && root !== null)) {
            return false;
        }
        const { a, b }: SafeShallowShape<Element> = root;
        if (!(typeof a === "number")) {
            return false;
        }
        if (!(typeof b === "string")) {
            return false;
        }
        ensureType<Element>({
            a,
            b
        });
        return true;
    }
    if (!(safeIsArray(root) && root.every(isElement))) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`nested object 1`] = `
"export function isX(root: unknown): root is X {
    if (!(typeof root === "object" && root !== null)) {
        return false;
    }
    const { a }: SafeShallowShape<X> = root;
    if (!(typeof a === "object" && a !== null)) {
        return false;
    }
    const { b }: SafeShallowShape<X["a"]> = a;
    if (!(typeof b === "object" && b !== null)) {
        return false;
    }
    const { c }: SafeShallowShape<X["a"]["b"]> = b;
    if (!(typeof c === "string")) {
        return false;
    }
    ensureType<X>({
        a: {
            b: {
                c
            }
        }
    });
    return true;
}
"
`;

exports[`null 1`] = `
"export function isX(root: unknown): root is X {
    if (!(root === null)) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`object with primitive types 1`] = `
"export function isX(root: unknown): root is X {
    if (!(typeof root === "object" && root !== null)) {
        return false;
    }
    const { a, b, c }: SafeShallowShape<X> = root;
    if (!(typeof a === "number")) {
        return false;
    }
    if (!(typeof b === "string")) {
        return false;
    }
    if (!(typeof c === "boolean")) {
        return false;
    }
    ensureType<X>({
        a,
        b,
        c
    });
    return true;
}
"
`;

exports[`reference type in a union 1`] = `
"export function isX(root: unknown): root is X {
    if (!((typeof root === "number") || (isA(root)))) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`reference type in an object 1`] = `
"export function isX(root: unknown): root is X {
    if (!(typeof root === "object" && root !== null)) {
        return false;
    }
    const { a, b }: SafeShallowShape<X> = root;
    if (!(isA(a))) {
        return false;
    }
    if (!(isB(b))) {
        return false;
    }
    ensureType<X>({
        a,
        b
    });
    return true;
}
"
`;

exports[`two arrays of a primitive type 1`] = `
"export function isX(root: unknown): root is X {
    type Element = X["a"][number];
    function isElement(root: unknown): root is Element {
        if (!(typeof root === "number")) {
            return false;
        }
        ensureType<Element>(root);
        return true;
    }
    type B_Element = X["b"][number];
    function isB_Element(root: unknown): root is B_Element {
        if (!(typeof root === "string")) {
            return false;
        }
        ensureType<B_Element>(root);
        return true;
    }
    if (!(typeof root === "object" && root !== null)) {
        return false;
    }
    const { a, b }: SafeShallowShape<X> = root;
    if (!(safeIsArray(a) && a.every(isElement))) {
        return false;
    }
    if (!(safeIsArray(b) && b.every(isB_Element))) {
        return false;
    }
    ensureType<X>({
        a,
        b
    });
    return true;
}
"
`;

exports[`undefined 1`] = `
"export function isX(root: unknown): root is X {
    if (!(root === undefined)) {
        return false;
    }
    ensureType<X>(root);
    return true;
}
"
`;

exports[`union of primitive and literal types 1`] = `
"export function isUnion(root: unknown): root is Union {
    if (!((typeof root === "string") || (typeof root === "number") || (root === null))) {
        return false;
    }
    ensureType<Union>(root);
    return true;
}
"
`;
